### PR TITLE
[5.6] Allow closure sub queries in wherePivot method

### PR DESCRIPTION
**Pull request by timacdonald**
Previous discussion: https://github.com/laravel/internals/issues/742

**TLDR;** This allows sub queries for `wherePivot` using a closure to avoid ambiguous column collisions.

---

There is an issue where working with pivot table queries can lead to ambiguous column names in sub queries unless the table name is prefixed.

This introduces the same closure based sub queries that `where` allows for `wherePivot`. This is nice as when calling `where` with a closure the query passed to the closure is no longer a `BelongsToMany` relation, thus the `wherePivot` method is not available within the sub query.

Previously:

``` php
$this->ceremonies()
     ->wherePivot('submissions_open_at', '<', Carbon::now())
     ->where(function ($query) {
         $query->where($this->ceremonies()->getTable().'.submissions_close_at', '>', Carbon::now())
             ->orWhere($this->ceremonies()->getTable().'.payments_close_at', '>', Carbon::now())
             ->orWhere($this->ceremonies()->getTable().'.uploads_close_at', '>', Carbon::now());
     });
```

With this PR:

``` php
$this->ceremonies()
     ->wherePivot('submissions_open_at', '<', Carbon::now())
     ->wherePivot(function ($query) {
         $query->wherePivot('submissions_close_at', '>', Carbon::now())
             ->orWherePivot('payments_close_at', '>', Carbon::now())
             ->orWherePivot('uploads_close_at', '>', Carbon::now());
     });
```

This will now prepend the column names with the pivot table name automatically.

### CODE REVIEW

**Comment by sisve** [comment]
There's a slight change in behavior; the code was previously returning whatever `$this->where(...)` returned, but is now changed into returning `$this`. I think it should keep the previous behavior, returning whatever `$this->where(...)` returns.

**Comment by timacdonald** [comment]
@sisve thanks for the feedback - good thinking. I've reverted back to my initial version which still `return $this->where(...)`.

**Comment by timacdonald** [comment]
I've found an issue with this. Closing until I can address it properly.

**Comment by timacdonald** [comment]
I've just added a test for this. Please let me know if it is not suitable or in the wrong location and I'll happily adjust.

I also found an issue in my original version, I've had to further reset the query that is passed to the closure to fix that.

**Comment by taylorotwell** [comment]
Can you explain what you are doing with the static constraints property?

**Comment by timacdonald** [comment]
@taylorotwell sure thing.

As we are passing to the closure a 'fresh' Relation query, I'm creating a instance in the `freshInstanceForSubPivotQuery` method. However, when a `BelongsToMany` is constructed - it has constraints added - specifically:

``` php
$this->query->where(
    $this->getQualifiedForeignPivotKeyName(), '=', $this->parent->{$this->parentKey}
);
```

This `where` constraint is not needed on a 'sub' query as it would have already been applied on the parent query. i.e. when you call

```
$user->roles(); // contraint is added.
```

So what I thought would be a good solution is to 'backup' and then 'restore' the value. I've just realised I should have set this to a `bool` though instead of `null`. I'll fix this now. Hopefully that all made sense.

**Comment by timacdonald** [comment]
Probably easier to just see the sql out put - rather than my rambling :)

This is the SQL from the test I wrote. The first one shows the SQL if I **don't** play with the constraints. The second is what happens with the code in this PR.

``` sql
// with constraints
"select * from "roles" inner join "eloquent_test_user_role" on "roles"."id" = "eloquent_test_user_role"."role_id" where "eloquent_test_user_role"."eloquent_test_user_id" is null and ("eloquent_test_user_role"."eloquent_test_user_id" is null and "eloquent_test_user_role"."pivot_table_column" = ?)"
// without constraints (my PR)
"select * from "roles" inner join "eloquent_test_user_role" on "roles"."id" = "eloquent_test_user_role"."role_id" where "eloquent_test_user_role"."eloquent_test_user_id" is null and ("eloquent_test_user_role"."pivot_table_column" = ?)"
```

**Comment by themsaid** [comment]
@timacdonald looks like your forked branch wasn't up to date with master, when I made a commit to your PR it looks like the commit history got broken, can you please open a new PR with these few changes?

```php
/**
 * Set a where clause for a pivot table column.
 *
 * @param  string  $column|\Closure
 * @param  string  $operator
 * @param  mixed   $value
 * @param  string  $boolean
 * @return \Illuminate\Database\Eloquent\Relations\BelongsToMany
 */
public function wherePivot($column, $operator = null, $value = null, $boolean = 'and')
{
    if ($column instanceof Closure) {
        $query = $this->getModel()->newQueryWithoutScopes();

        call_user_func($column, $this->freshInstanceForSubPivotQuery($query));

        $this->query->addNestedWhereQuery($query->getQuery(), $boolean);

        return $this;
    }

    $this->pivotWheres[] = func_get_args();

    return $this->where($this->table.'.'.$column, $operator, $value, $boolean);
}

/**
 * Create a fresh instance with query.
 *
 * @param  \Illuminate\Database\Eloquent\Builder  $query
 * @return static
 */
protected function freshInstanceForSubPivotQuery($query)
{
    $instance =  $this->noConstraints(function () use ($query) {
        return new static(
            $query, $this->parent, $this->table, $this->foreignPivotKey,
            $this->relatedPivotKey, $this->parentKey, $this->relatedKey,
            $this->relationName
        );
    });

    return tap($instance, function ($instance) {
        $instance->getBaseQuery()->joins = null;
    });
}
```

**Comment by timacdonald** [comment]
Ah dang your totally right I was gonna target 5.5 I'll get that going for u

**Comment by MattWohler** [comment]
> Why was it deleted? This helps for IDE
> If this is a conflict, is the new version still being used?

@nurmanhabib , The comment before yours shows him closing this one and opening a new PR, however shutting it down due to new(unstable) issues that arose.
